### PR TITLE
Added unitest for support-package

### PR DESCRIPTION
**Pull request by Raghul-M**
# Issue link
Closes https://github.com/project-codeflare/codeflare-operator/issues/239

# What changes have been made
Added Unit test for the Support Package Applicable Modules 
Added Workflow file to test the Unit tests

# Verification steps
RUN - `go test  ./support -v`

## Checks
- [x] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [ ] Manual tests
   - [ ] Testing is not required for this change

### CODE REVIEW

File: support/batch_test.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/onsi/gomega"
+
+	batchv1 "k8s.io/api/batch/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func TestGetJob(t *testing.T) {
+
+	g := gomega.NewGomegaWithT(t)
```
Comment (sutaakar) [inline support/batch_test.go:31]: Do you need gomega in here? IMHO all usage can be replaced by `test` which you declare below.

File: support/batch_test.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	fakeClient := NewFakeKubeClientWithObjects(fakeJobs...)
+
+	test := With(t).(*T)
+	test.client = &testClient{
+		core: fakeClient,
+	}
+
+	// Call the Job function using the fake client
+	jobFunc := Job(test, "my-namespace", "my-job-1")
+	job := jobFunc(g)
```
Comment (sutaakar) [inline support/batch_test.go:50]:
```suggestion
	job := Job(test, "my-namespace", "my-job-1")(g)
```

File: support/core_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func TestGetPods(t *testing.T) {
+	// Create a fake Kubernetes client for testing
+	fakeClient := fake.NewSimpleClientset(&corev1.Pod{
```
Comment (sutaakar) [inline support/core_test.go:15]: This can be replaced by `NewFakeKubeClientWithObjects`, right?

File: support/core_test.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	// Call the GetPods function with the fake client and namespace
+	pods := GetPods(test, "test-namespace", metav1.ListOptions{})
+
+	test.Expect(pods).Should(gomega.HaveLen(1), "Expected 1 pod, but got %d", len(pods))
+	test.Expect(pods[0].Name).To(gomega.Equal("test-pod"), "Expected pod name 'test-pod', but got '%s'", pods[0].Name)
+}
+
+func TestGetNodes(t *testing.T) {
+	// Create a fake Kubernetes client for testing
+	fakeClient := fake.NewSimpleClientset(&corev1.Node{
```
Comment (sutaakar) [inline support/core_test.go:36]: Same here

File: support/environment_test.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+func TestGetCodeFlareSDKVersion(t *testing.T) {
+	// Set the environment variable.
+	os.Setenv(CodeFlareTestSdkVersion, "1.2.3")
+
+	// Get the version.
+	version := GetCodeFlareSDKVersion()
+
+	// Assert that the version is correct.
+	if version != "1.2.3" {
```
Comment (sutaakar) [inline support/environment_test.go:19]: The condition here seems to be not necessary

File: support/environment_test.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+func TestGetRayVersion(t *testing.T) {
+	// Set the environment variable.
+	os.Setenv(CodeFlareTestRayVersion, "1.4.5")
+
+	// Get the version.
+	version := GetRayVersion()
+
+	// Assert that the version is correct.
+	if version != "1.4.5" {
```
Comment (sutaakar) [inline support/environment_test.go:33]: Same here

File: support/batch_test.go (hunk 1 above)
Comment (Raghul-M) [reply support/batch_test.go]: Gotcha we don't need it. Yes I replace it with test as well as changed in ingress_test.go

File: support/environment_test.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+func TestGetRayImage(t *testing.T) {
+	// Set the environment variable.
+	os.Setenv(CodeFlareTestRayImage, "ray/ray:latest")
+
+	// Get the image.
+	image := GetRayImage()
+
+	// Assert that the image is correct.
+	if image != "ray/ray:latest" {
```
Comment (sutaakar) [inline support/environment_test.go:46]: And here

File: support/environment_test.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+func TestGetPyTorchImage(t *testing.T) {
+	// Set the environment variable.
+	os.Setenv(CodeFlareTestPyTorchImage, "pytorch/pytorch:latest")
+
+	// Get the image.
+	image := GetPyTorchImage()
+
+	// Assert that the image is correct.
+	if image != "pytorch/pytorch:latest" {
```
Comment (sutaakar) [inline support/environment_test.go:60]: And here

File: support/environment_test.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func TestGetClusterID(t *testing.T) {
+	os.Setenv(ClusterID, "my-cluster-id")
+	clusterId, ok := GetClusterId()
+	if !ok {
+		gomega.Expect(ok).To(gomega.BeTrue(), "Expected GetClusterId() to return true, but got false.")
+	}
+	if clusterId != "my-cluster-id" {
+		gomega.Expect(clusterId).To(gomega.Equal("my-cluster-id"), "Expected GetClusterId() to return 'my-cluster-id', but got '%s'.", clusterId)
```
Comment (sutaakar) [inline support/environment_test.go:73]: And here

File: support/environment_test.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func TestGetInstascaleOcmSecret(t *testing.T) {
+	// Set the Instascale OCM secret environment variable.
+	os.Setenv(InstaScaleOcmSecret, "default/instascale-ocm-secret")
+	// Get the Instascale OCM secret namespace and secret name.
+	namespace, secretName := GetInstascaleOcmSecret()
+
+	// Verify that the namespace and secret name are correct.
+	if namespace != "default" || secretName != "instascale-ocm-secret" {
```
Comment (sutaakar) [inline support/environment_test.go:84]: And here

File: support/environment_test.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+			envVarValue: "KIND",
+			expected:    KindCluster,
+		},
+	}
+	ttt := With(t)
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			os.Setenv(ClusterTypeEnvVar, tt.envVarValue)
+			actual := GetClusterType(ttt)
+			if actual != tt.expected {
```
Comment (sutaakar) [inline support/environment_test.go:127]: And here

File: support/core_test.go (hunk 3 above)
Comment (Raghul-M) [reply support/core_test.go:15]: Right Replaced

File: support/core_test.go (hunk 4 above)
Comment (Raghul-M) [reply support/core_test.go:36]: Done

File: support/fakeclientwithscheme.go [hunk 12]
```diff
@@ -0,0 +1,11 @@
+package support
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func NewFakeKubeClientWithScheme(scheme *runtime.Scheme, objects ...client.Object) client.Client {
```
Comment (sutaakar) [inline support/fakeclientwithscheme.go:9]: Can you move it to `fakeclient.go`. There is no need to have separate files for similar functions.

File: support/image_test.go [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+				Name:      "my-imagestream-1",
+				Namespace: "my-namespace",
+			},
+		},
+	}
+	fakeClient := NewFakeKubeClientWithScheme(scheme, fakeImageStream...)
+
+	image := &imagev1.ImageStream{}
+
+	err := fakeClient.Get(context.TODO(), client.ObjectKey{Name: "my-imagestream-1", Namespace: "my-namespace"}, image)
```
Comment (sutaakar) [inline support/image_test.go:34]: This test should call `GetImageStream` function.

File: support/ingress_test.go [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+	fakeClient := NewFakeKubeClientWithObjects(fakeIngress...)
+
+	test := With(t).(*T)
+	test.client = &testClient{
+		core: fakeClient,
+	}
+
+	// Call the Ingress function using the fake client
+	ingressFunc := Ingress(test, "my-namespace", "my-ingress-1")
+	ingress := ingressFunc(test)
```
Comment (sutaakar) [inline support/ingress_test.go:34]:
```suggestion
	ingress := Ingress(test, "my-namespace", "my-ingress-1")(test)
```

File: support/machine_test.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-machineset-1",
+				Namespace: "openshift-machine-api",
+			},
+		},
+	}
+	fakeClient := NewFakeKubeClientWithScheme(scheme, testmachines...)
+
+	machine := &machinev1beta1.MachineSet{}
+	err := fakeClient.Get(context.TODO(), client.ObjectKey{Name: "test-machineset-1", Namespace: "openshift-machine-api"}, machine)
```
Comment (sutaakar) [inline support/machine_test.go:33]: `GetMachineSets` should be called here.

File: support/ray_test.go [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+				Name:      "my-job-1",
+				Namespace: "my-namespace",
+			},
+		},
+	}
+
+	fakeClient := NewFakeKubeClientWithScheme(scheme, fakeRayJobs...)
+
+	rayJob := &rayv1alpha1.RayJob{}
+	err := fakeClient.Get(context.TODO(), client.ObjectKey{Name: "my-job-1", Namespace: "my-namespace"}, rayJob)
```
Comment (sutaakar) [inline support/ray_test.go:33]: Same here

File: support/ray_test.go [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+				Name:      "my-cluster-1",
+				Namespace: "my-namespace",
+			},
+		},
+	}
+
+	fakeClient := NewFakeKubeClientWithScheme(scheme, fakeRayCluster...)
+
+	raycluster := &rayv1alpha1.RayCluster{}
+	err := fakeClient.Get(context.TODO(), client.ObjectKey{Name: "my-cluster-1", Namespace: "my-namespace"}, raycluster)
```
Comment (sutaakar) [inline support/ray_test.go:58]: And here

File: support/route_test.go [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-1",
+				Namespace: "my-namespace",
+			},
+		},
+	}
+	fakeClient := NewFakeKubeClientWithScheme(scheme, fakeroute...)
+
+	route := &routev1.Route{}
+	err := fakeClient.Get(context.TODO(), client.ObjectKey{Name: "test-1", Namespace: "my-namespace"}, route)
```
Comment (sutaakar) [inline support/route_test.go:34]: And here

File: support/fakeclient.go [hunk 19]
```diff
@@ -0,0 +1,11 @@
+package support
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func NewFakeKubeClientWithObjects(objects ...runtime.Object) *fake.Clientset {
```
Comment (sutaakar) [inline support/fakeclient.go:8]: I see a common usage to initialize client using for example:
```
	fakeClient := NewFakeKubeClientWithObjects(fakeIngress...)

	test := With(t).(*T)
	test.client = &testClient{
		core: fakeClient,
	}
```

Wouldn't it have sense to add it here and return `test`?

File: support/ingress_test.go (hunk 14 above)
Comment (Raghul-M) [reply support/ingress_test.go:34]: Done, Thanks for the suggestion

File: support/fakeclientwithscheme.go (hunk 12 above)
Comment (Raghul-M) [reply support/fakeclientwithscheme.go:9]: Done

File: support/image_test.go (hunk 13 above)
Comment (Raghul-M) [reply support/image_test.go:34]: Resolved

File: support/ray_test.go (hunk 17 above)
Comment (Raghul-M) [reply support/ray_test.go:58]: Resolved

File: support/route_test.go (hunk 18 above)
Comment (Raghul-M) [reply support/route_test.go:34]: Resolved

File: support/ray_test.go (hunk 16 above)
Comment (Raghul-M) [reply support/ray_test.go:33]: Resolved

File: support/fakeclient.go (hunk 19 above)
Comment (Raghul-M) [reply support/fakeclient.go:8]: Resolved as per suggestion

File: support/machine_test.go [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+	}
+
+	test.client.Machine().MachineV1beta1().MachineSets("openshift-machine-api").Create(test.ctx, machine, metav1.CreateOptions{})
+
+	machines, _ := GetMachineSets(test)
+
+	test.Expect(machines[0].Name).To(gomega.Equal("test-machineset-1"))
+	test.Expect(machines[0].Namespace).To(gomega.Equal("openshift-machine-api"))
+	//test.Expect(machines{client: nil, ns: "test-machineset-1"}).To(gomega.Equal(machineSets{client: nil, ns: "test-machineset-1"}))
+	//test.Expect(machines{client: nil, ns: "openshift-machine-api"}).To(gomega.Equal(machineSets{client: nil, ns: "openshift-machine-api"}))
```
Comment (sutaakar) [inline support/machine_test.go:30]: These lines can be removed.

File: support/machine_test.go [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+			Name:      "test-machineset-1",
+			Namespace: "openshift-machine-api",
+		},
+	}
+
+	test.client.Machine().MachineV1beta1().MachineSets("openshift-machine-api").Create(test.ctx, machine, metav1.CreateOptions{})
+
+	machines, _ := GetMachineSets(test)
+
+	test.Expect(machines[0].Name).To(gomega.Equal("test-machineset-1"))
```
Comment (sutaakar) [inline support/machine_test.go:27]: Can you add assertion making sure there is just one element in `machines` slice?

File: support/machine_test.go (hunk 20 above)
Comment (Raghul-M) [reply support/machine_test.go]: Removed

File: support/machine_test.go (hunk 21 above)
Comment (Raghul-M) [reply support/machine_test.go:26]: Assertion Added

File: support/fakeclient.go [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+
+	"k8s.io/apimachinery/pkg/runtime"
+	fakeCore "k8s.io/client-go/kubernetes/fake"
+
+	fakeimage "github.com/openshift/client-go/image/clientset/versioned/fake"
+	fakeMachine "github.com/openshift/client-go/machine/clientset/versioned/fake"
+	fakeroute "github.com/openshift/client-go/route/clientset/versioned/fake"
+)
+
+func NewTest(t *testing.T, objects ...runtime.Object) *T {
```
Comment (sutaakar) [inline support/fakeclient.go:16]:
```suggestion
func NewTest(t *testing.T) *T {
```

Comment (Raghul-M) [reply support/fakeclient.go]: Changed

**Review by sutaakar — APPROVED** [review]
/lgtm

**Pull request merged**